File: packages/altimate-engine/src/altimate_engine/docker_discovery.py
from __future__ import annotations

from typing import Any

# Only include DB types the engine has connectors for
IMAGE_MAP = {
    "postgres": {
        "type": "postgres",
        "port": 5432,
        "env_user": "POSTGRES_USER",
        "env_password": "POSTGRES_PASSWORD",
        "env_database": "POSTGRES_DB",
    },
    "mysql": {
        "type": "mysql",
        "port": 3306,
        "env_user": "MYSQL_USER",
        "env_password": "MYSQL_PASSWORD",
        "env_database": "MYSQL_DATABASE",
        "alt_password": "MYSQL_ROOT_PASSWORD",
    },
    "mariadb": {
        "type": "mysql",
        "port": 3306,
        "env_user": "MARIADB_USER",
        "env_password": "MARIADB_PASSWORD",
        "env_database": "MARIADB_DATABASE",
        "alt_password": "MARIADB_ROOT_PASSWORD",
    },
    "mcr.microsoft.com/mssql": {
        "type": "sqlserver",
        "port": 1433,
        "env_password": "SA_PASSWORD",
    },
}


def _match_image(image: str) -> dict[str, Any] | None:
    image_lower = image.lower()
    for pattern, config in IMAGE_MAP.items():
        if pattern in image_lower:
            return config
    return None


def _extract_port(container: Any, default_port: int) -> int | None:
    ports = container.attrs.get("NetworkSettings", {}).get("Ports", {})
    for port_key, mappings in ports.items():
        if mappings:
            host_port = mappings[0].get("HostPort")
            if host_port:
                return int(host_port)
    return None


def discover_containers() -> list[dict[str, Any]]:
    try:
        import docker
    except ImportError:
        return []

    try:
        client = docker.from_env()
    except Exception:
        return []

    results = []

    for container in client.containers.list():
        try:
            image_name = container.attrs.get("Config", {}).get("Image", "")
            image_config = _match_image(image_name)
            if not image_config:
                continue

            env_vars = {}
            for env in container.attrs.get("Config", {}).get("Env", []):
                if "=" in env:
                    key, value = env.split("=", 1)
                    env_vars[key] = value

            port = _extract_port(container, image_config["port"])
            if port is None:
                continue

            conn: dict[str, Any] = {
                "container_id": container.id[:12],
                "name": container.name,
                "image": image_name,
                "db_type": image_config["type"],
                "host": "localhost",
                "port": port,
                "status": container.status,
            }

            if "env_user" in image_config and image_config["env_user"] in env_vars:
                conn["user"] = env_vars[image_config["env_user"]]
            if (
                "env_password" in image_config
                and image_config["env_password"] in env_vars
            ):
                conn["password"] = env_vars[image_config["env_password"]]
            elif (
                "alt_password" in image_config
                and image_config["alt_password"] in env_vars
            ):
                conn["password"] = env_vars[image_config["alt_password"]]
            if (
                "env_database" in image_config
                and image_config["env_database"] in env_vars
            ):
                conn["database"] = env_vars[image_config["env_database"]]

            results.append(conn)
        except Exception:
            continue

    return results

File: packages/opencode/src/altimate/native/connections/docker-discovery.ts
/**
 * Docker container detection for database connections.
 *
 * Uses the `dockerode` package (dynamic import). Returns empty array
 * if dockerode is not installed or Docker is not running.
 */

import type { ConnectionConfig } from "@altimateai/drivers"
import type { DockerContainer } from "../types"

/** Map container images to database types. */
const IMAGE_MAP: Array<{ pattern: RegExp; type: string }> = [
  { pattern: /postgres/i, type: "postgres" },
  { pattern: /mysql/i, type: "mysql" },
  { pattern: /mariadb/i, type: "mysql" },
  { pattern: /mcr\.microsoft\.com\/mssql/i, type: "sqlserver" },
  { pattern: /mssql/i, type: "sqlserver" },
  { pattern: /oracle/i, type: "oracle" },
  { pattern: /gvenzl\/oracle/i, type: "oracle" },
  { pattern: /clickhouse/i, type: "clickhouse" },
]

/** Map environment variable names to connection config fields by db type. */
const ENV_MAP: Record<string, Record<string, string>> = {
  postgres: {
    POSTGRES_USER: "user",
    POSTGRES_PASSWORD: "password",
    POSTGRES_DB: "database",
  },
  mysql: {
    MYSQL_USER: "user",
    MYSQL_ROOT_PASSWORD: "password",
    MYSQL_PASSWORD: "password",
    MYSQL_DATABASE: "database",
  },
  sqlserver: {
    SA_PASSWORD: "password",
    MSSQL_SA_PASSWORD: "password",
  },
  oracle: {
    ORACLE_PASSWORD: "password",
    APP_USER: "user",
    APP_USER_PASSWORD: "password",
    ORACLE_DATABASE: "database",
  },
  clickhouse: {
    CLICKHOUSE_USER: "user",
    CLICKHOUSE_PASSWORD: "password",
    CLICKHOUSE_DB: "database",
  },
}

/** Default ports by database type. */
const DEFAULT_PORTS: Record<string, number> = {
  postgres: 5432,
  mysql: 3306,
  sqlserver: 1433,
  oracle: 1521,
  clickhouse: 8123,
}

/** Default users by database type. */
const DEFAULT_USERS: Record<string, string> = {
  postgres: "postgres",
  mysql: "root",
  sqlserver: "sa",
  oracle: "system",
  clickhouse: "default",
}

function detectDbType(image: string): string | null {
  for (const { pattern, type } of IMAGE_MAP) {
    if (pattern.test(image)) return type
  }
  return null
}

function parseEnvVars(envList: string[], dbType: string): Record<string, string> {
  const result: Record<string, string> = {}
  const mapping = ENV_MAP[dbType] ?? {}

  for (const env of envList) {
    const eqIdx = env.indexOf("=")
    if (eqIdx < 0) continue
    const key = env.slice(0, eqIdx)
    const value = env.slice(eqIdx + 1)
    const configField = mapping[key]
    if (configField && !result[configField]) {
      result[configField] = value
    }
  }

  return result
}

function extractPort(ports: Record<string, any>[] | undefined, dbType: string): number {
  const defaultPort = DEFAULT_PORTS[dbType] ?? 5432
  if (!ports || !Array.isArray(ports)) return defaultPort

  for (const p of ports) {
    if (p.PublicPort && p.PrivatePort === defaultPort) {
      return p.PublicPort
    }
  }

  // Fall back to first public port
  for (const p of ports) {
    if (p.PublicPort) return p.PublicPort
  }

  return defaultPort
}

/**
 * Discover database containers running in Docker.
 * Returns an array of DockerContainer descriptions.
 */
export async function discoverContainers(): Promise<DockerContainer[]> {
  let Docker: any
  try {
    // @ts-expect-error — optional dependency, loaded at runtime
    const mod = await import("dockerode")
    Docker = mod.default || mod
  } catch {
    return []
  }

  try {
    const docker = new Docker()
    const containers = await docker.listContainers({ all: false })
    const results: DockerContainer[] = []

    for (const container of containers) {
      const image = container.Image ?? ""
      const dbType = detectDbType(image)
      if (!dbType) continue

      const envVars = parseEnvVars(container.Env ?? [], dbType)
      const port = extractPort(container.Ports, dbType)
      const name = (container.Names?.[0] ?? "").replace(/^\//, "")

      results.push({
        container_id: (container.Id ?? "").slice(0, 12),
        name,
        image,
        db_type: dbType,
        host: "127.0.0.1",
        port,
        user: envVars.user ?? DEFAULT_USERS[dbType],
        password: envVars.password,
        database: envVars.database,
        status: container.State ?? container.Status ?? "unknown",
      })
    }

    return results
  } catch {
    // Docker not running or permission error
    return []
  }
}

/**
 * Convert a discovered Docker container to a ConnectionConfig.
 */
export function containerToConfig(container: DockerContainer): ConnectionConfig {
  const config: ConnectionConfig = {
    type: container.db_type,
    host: container.host,
    port: container.port,
  }
  if (container.user) config.user = container.user
  if (container.password) config.password = container.password
  if (container.database) config.database = container.database
  return config
}
